driver-core/src/main/com/mongodb/operation/AsyncQueryBatchCursor.java

```diff
@@ -108,7 +110,17 @@
 
     @Override
     public void close() {
-        if (!isClosed.getAndSet(true)) {
+        boolean closed;
```

Comment:
I find this logic hard to follow.  I think it can be simplified to something like this:

        boolean killCursor = false;
        synchronized (this) {
            if (isOperationInProgress) {
                isClosePending = true;
            }
            else {
                // NOTE: only attempt to kill cursor on first close
                killCursor = !isClosed;
                isClosed = true;
            }
        }

        if (killCursor) {
            killCursorOnClose();
        }

(this assumes the changes I requested to the fields have been applied)

Does that look correct, and also simpler, to you?

Reply:
Yes, it is more clear. I have incorporated the code.